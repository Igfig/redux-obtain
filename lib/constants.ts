export const REDUX_STORE_NAME = "resources"

export const initialState = {}
export const initialResourceState = {
    data: {},
    loading: false,
    error: false
}

export const ADD_RESOURCE = "@@fetcher/ADD_RESOURCE"
export const MODIFY_RESOURCE = "@@fetcher/MODIFY_RESOURCE"
export const REQUEST_RESOURCE = "@@fetcher/REQUEST_RESOURCE"
export const FETCH_SUCCESS = "@@fetcher/FETCH_SUCCESS"
export const FETCH_ADDITIONAL_SUCCESS = "@@fetcher/FETCH_ADDITIONAL_SUCCESS"
export const FETCH_ERROR = "@@fetcher/FETCH_ERROR"
export const REMOVE_RESOURCE = "@@fetcher/REMOVE_RESOURCE"

export const LIST_FILTER_VALUES = [
    "pcps",
    "units",
    "payers",
    "productClasses",
    "acuteEncounters",
    "measures",
    "conditions"
]
