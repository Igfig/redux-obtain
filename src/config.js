export let config = {
    requestHeaderSelector: state => ({
        "Content-Type": "application/json",
        Authorization: `Basic ${state.authentication.token}`
    }),
    reduxStoreName: "resources",
    paginationInitialLoadLimit: 500,
    getOrderBys: props =>
        props && props.sortBy && props.sortDirection
            ? props.sortBy.map((column, index) => ({
                  column,
                  direction: props.sortDirection[index]
              }))
            : []
}

export const setupFetcher = options => {
    config = {
        ...config,
        ...options
    }
}
