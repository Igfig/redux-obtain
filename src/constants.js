export const initialState = {}
export const initialResourceState = {
    data: {},
    loading: false,
    error: false
}

const prefix = "@@fetcher"

export const ADD_RESOURCE = `${prefix}/ADD_RESOURCE`
export const MODIFY_RESOURCE = `${prefix}/MODIFY_RESOURCE`
export const REQUEST_RESOURCE = `${prefix}/REQUEST_RESOURCE`
export const FETCH_SUCCESS = `${prefix}/FETCH_SUCCESS`
export const FETCH_ADDITIONAL_SUCCESS = `${prefix}/FETCH_ADDITIONAL_SUCCESS`
export const FETCH_ERROR = `${prefix}/FETCH_ERROR`
export const REMOVE_RESOURCE = `${prefix}/REMOVE_RESOURCE`
