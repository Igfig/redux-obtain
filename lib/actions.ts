import * as C from "./constants"

export const addResource = (name: string, paginationKey?: string) => ({
    type: C.ADD_RESOURCE,
    name,
    paginationKey
})

export const modifyResource = ({ name, dataTransform }: any) => ({
    type: C.MODIFY_RESOURCE,
    name,
    dataTransform
})

export const requestResource = (name: string) => ({ type: C.REQUEST_RESOURCE, name })

export const fetchSuccess = (name: string, data: any, acceptResponse: any) => ({
    type: C.FETCH_SUCCESS,
    name,
    data,
    acceptResponse
})

export const fetchAdditionalSuccess = (name: string, data: any) => ({
    type: C.FETCH_ADDITIONAL_SUCCESS,
    name,
    data
})

export const fetchError = (name: any, error: any) => ({ type: C.FETCH_ERROR, name, error })

export const removeResource = (name: any) => ({ type: C.REMOVE_RESOURCE, name })
