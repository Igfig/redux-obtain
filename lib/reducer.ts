import * as C from "./constants"

const createReducer = (initialState: any, handlers: any) => (state = initialState, action: any) =>
    handlers.hasOwnProperty(action.type) ? handlers[action.type](state, action) : state

export const reducer = createReducer(C.initialState, {
    [C.ADD_RESOURCE]: (state: any, action: any) => ({
        ...state,
        [action.name]: {
            ...C.initialResourceState,
            paginationKey: action.paginationKey
        }
    }),
    [C.MODIFY_RESOURCE]: (state: any, action: any) => ({
        ...state,
        [action.name]: state[action.name] && {
            ...state[action.name],
            data: action.dataTransform(state[action.name] && state[action.name].data)
        }
    }),
    [C.REQUEST_RESOURCE]: (state: any, action: any) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            data: {},
            loading: true,
            error: false
        }
    }),
    [C.FETCH_SUCCESS]: (state: any, action: any) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            data: action.acceptResponse ? action.acceptResponse(action.data) : action.data,
            loading: false,
            error: false
        }
    }),
    [C.FETCH_ADDITIONAL_SUCCESS]: (state: any, action: any) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            data: {
                ...action.data,
                [state[action.name].paginationKey]: state[action.name].paginationKey && [
                    ...state[action.name].data[state[action.name].paginationKey],
                    ...action.data[state[action.name].paginationKey]
                ]
            },
            loading: false,
            error: false
        }
    }),
    [C.FETCH_ERROR]: (state: any, action: any) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            data: {},
            loading: false,
            error: true
        }
    }),
    [C.REMOVE_RESOURCE]: (state: any, action: any) => ({
        ...state,
        [action.name]: undefined
    })
})
