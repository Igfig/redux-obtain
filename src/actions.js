import * as C from "./constants"

export const addResource = (name, paginationKey) => ({
    type: C.ADD_RESOURCE,
    name,
    paginationKey
})

export const modifyResource = ({ name, dataTransform }) => ({
    type: C.MODIFY_RESOURCE,
    name,
    dataTransform
})

export const requestResource = name => ({ type: C.REQUEST_RESOURCE, name })

export const fetchSuccess = (name, data, acceptResponse) => ({
    type: C.FETCH_SUCCESS,
    name,
    data,
    acceptResponse
})

export const fetchAdditionalSuccess = (name, data) => ({
    type: C.FETCH_ADDITIONAL_SUCCESS,
    name,
    data
})

export const fetchError = (name, error) => ({ type: C.FETCH_ERROR, name, error })

export const removeResource = name => ({ type: C.REMOVE_RESOURCE, name })
