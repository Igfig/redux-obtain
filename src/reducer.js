import * as C from "./constants"

const createReducer = (initialState, handlers) => (state = initialState, action) =>
    handlers.hasOwnProperty(action.type) ? handlers[action.type](state, action) : state

export const reducer = createReducer(C.initialState, {
    [C.ADD_RESOURCE]: (state, action) => ({
        ...state,
        [action.name]: {
            ...C.initialResourceState,
            paginationKey: action.paginationKey
        }
    }),
    [C.MODIFY_RESOURCE]: (state, action) => ({
        ...state,
        [action.name]: state[action.name] && {
            ...state[action.name],
            data: action.dataTransform(state[action.name] && state[action.name].data)
        }
    }),
    [C.REQUEST_RESOURCE]: (state, action) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            loading: true,
            error: false
        }
    }),
    [C.FETCH_SUCCESS]: (state, action) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            data: action.acceptResponse ? action.acceptResponse(action.data) : action.data,
            loading: false,
            error: false
        }
    }),
    [C.FETCH_ADDITIONAL_SUCCESS]: (state, action) => {
        const paginationKey = state[action.name].paginationKey
        const newData = action.acceptResponse ? action.acceptResponse(action.data) : action.data
        return {
            ...state,
            [action.name]: {
                ...state[action.name],
                data: {
                    ...newData,
                    [paginationKey]: paginationKey && [
                        ...state[action.name].data[paginationKey],
                        ...action.data[paginationKey]
                    ]
                },
                loading: false,
                error: false
            }
        }
    },
    [C.FETCH_ERROR]: (state, action) => ({
        ...state,
        [action.name]: {
            ...state[action.name],
            data: {},
            loading: false,
            error: true
        }
    }),
    [C.REMOVE_RESOURCE]: (state, action) => ({
        ...state,
        [action.name]: undefined
    })
})
