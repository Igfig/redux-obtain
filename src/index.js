import { fetcher } from "./fetcher"
import { modifyResource } from "./actions"
import { getResourceData, getMetaData } from "./selectors"
import { reducer } from "./reducer"
import { setupFetcher } from "./config"

export { fetcher, modifyResource, getResourceData, getMetaData, reducer, setupFetcher }
