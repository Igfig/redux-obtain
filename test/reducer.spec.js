import { reducer } from "../src/reducer"
import * as C from "../src/constants"
import * as actions from "../src/actions"
import {
    modifyResource,
    requestResource,
    fetchSuccess,
    fetchAdditionalSuccess,
    addResource,
    fetchError,
    removeResource
} from "../src/actions"

describe("Reducer", () => {
    describe("init", () => {
        it("should return initial state", () => {
            expect(reducer(undefined, {})).toMatchSnapshot()
        })
    })
    describe("Add Resource", () => {
        it("should add a resource", () => {
            expect(reducer(C.initialState, actions.addResource("new_resource"))).toMatchSnapshot()
        })
        it("should add a multiple resources", () => {
            const first_resource = "first_resource_name"
            const second_resource = "second_resource_name"
            const state1 = C.initialState
            const state2 = reducer(state1, actions.addResource(first_resource))
            const final_state = reducer(state2, actions.addResource(second_resource))
            expect(final_state).toMatchSnapshot()
        })
        it("should add a pagination key", () => {
            const pName = "paginated_resource"
            const pKey = "name_of_attribute_to_be_paginated_on"
            expect(reducer(C.initialState, actions.addResource(pName, pKey))).toMatchSnapshot()
        })
    })
    describe("Modify Resource", () => {
        it("should do nothing if resource does not exist", () => {
            expect(
                reducer(
                    C.initialState,
                    modifyResource({
                        name: "some_resource_name",
                        dataTransform: data => ({ ...data, list: [1, 2, 3, 4] })
                    })
                )
            ).toMatchSnapshot()
        })
        it("should do nothing if no dataTransform is given", () => {
            const name = "whatever"
            const state = reducer(C.initialState, actions.addResource(name))
            expect(reducer({ ...state }, modifyResource({ name }))).toMatchSnapshot()
        })
        it("should apply dataTransform", () => {
            const name = "whatever"
            const state = reducer(C.initialState, actions.addResource(name))
            expect(
                reducer(
                    { ...state },
                    modifyResource({
                        name,
                        dataTransform: data => ({ ...data, newKey: "newValue" })
                    })
                )
            ).toMatchSnapshot()
        })
    })
    describe("Request Resource", () => {
        it("should do nothing if the resource does not exist", () => {
            const name = "some name"
            expect(reducer(C.initialState, requestResource(name))).toMatchSnapshot()
        })
        it("should set flags on request", () => {
            const name = "random name"
            const state = {
                [name]: C.initialResourceState
            }
            expect(reducer(state, requestResource(name))).toMatchSnapshot()
        })
    })
    describe("Fetch Success", () => {
        it("should do nothing if the resource doesn't exist", () => {
            const name = "some name"
            expect(
                reducer(C.initialState, fetchSuccess(name, { list: [1, 2, 3, 4] }, data => data))
            ).toMatchSnapshot()
        })
        it("should set the flags correctly", () => {
            const name = "resource name"
            const data = { list: [1, 2, 3, 4] }
            expect(
                reducer({ [name]: C.initialResourceState }, fetchSuccess(name, data))
            ).toMatchSnapshot()
        })
        it("should set data without dataTransform", () => {
            const name = "resource name"
            const data = {
                list: [1, 2, 3, 4],
                complicated: { nested: ["structure", "and", "a", "half"] }
            }
            expect(
                reducer({ [name]: C.initialResourceState }, fetchSuccess(name, data))
            ).toMatchSnapshot()
        })
        it("should apply acceptResponse to data", () => {
            const name = "something something name"
            const data = {
                list: [1, 2, 3, 4, 5, 66],
                object: { foo: "bar" }
            }
            const acceptResponse = data => ({ ...data, object: { ...data.object, foo: "baz" } })
            expect(
                reducer(
                    { [name]: C.initialResourceState },
                    fetchSuccess(name, data, acceptResponse)
                )[name].data
            ).toMatchSnapshot()
        })
    })
    describe("Fetch Additional Success", () => {
        it("should do nothing if the resource does not exist", () => {
            expect(
                reducer(
                    C.initialState,
                    fetchAdditionalSuccess("some name", { list: [1, 2, 3, 4] }, data => data)
                )
            ).toMatchSnapshot()
        })
        it("should set the flags correctly", () => {
            const name = "eh whatever"
            expect(
                reducer(
                    {
                        [name]: {
                            ...C.initialResourceState,
                            data: { pKey: [] },
                            paginationKey: "pKey"
                        }
                    },
                    fetchAdditionalSuccess(name, { pKey: [1, 2, 3] })
                )
            ).toMatchSnapshot()
        })
        it("should append new data onto paginationKey", () => {
            const name = "resource name"
            const pKey = "pagination key"
            const state1 = reducer(C.initialState, addResource(name, pKey))
            const state2 = reducer(state1, fetchSuccess(name, { [pKey]: [1, 2, 3, 4] }))
            expect(
                reducer(state2, fetchAdditionalSuccess(name, { [pKey]: [5, 6, 7, 8] }))
            ).toMatchSnapshot()
        })
        it("should apply accept response to new data", () => {
            const name = "resource name"
            const pKey = "pagination key"
            const acceptResponse = data => ({ [pKey]: data[pKey].map(x => x * 2), plus: "extra" })
            const state1 = reducer(C.initialState, addResource(name, pKey))
            const state2 = reducer(state1, fetchSuccess(name, { [pKey]: [1, 2, 3, 4] }))
            expect(
                reducer(
                    state2,
                    fetchAdditionalSuccess(name, { [pKey]: [5, 6, 7, 8] }, acceptResponse)
                )
            ).toMatchSnapshot()
        })
    })
    describe("Fetch Error", () => {
        it("should do nothing if resource does not exist", () => {
            const name = "unlucky resource"
            expect(reducer(C.initialState, fetchError(name))).toMatchSnapshot()
        })
        it("should set flags and clear data", () => {
            const name = "unfortunate resource"
            expect(
                reducer(
                    { [name]: { ...C.initialResourceState, data: { something: "is here" } } },
                    fetchError(name)
                )
            ).toMatchSnapshot()
        })
    })
    describe("Remove Resource", () => {
        it("should do nothing if resource does not exist", () => {
            const name = "unknown resource"
            expect(reducer(C.initialState, removeResource(name))).toMatchSnapshot()
        })
        it("should remove the correct resource", () => {
            const r1 = "resource_1"
            const r2 = "resource_2"
            const state1 = reducer(C.initialState, addResource(r1))
            const state2 = reducer(state1, addResource(r2))
            expect(reducer(state2, removeResource(r2))).toMatchSnapshot()
            expect(reducer(state1, removeResource(r1))).toMatchSnapshot()
        })
    })
})
