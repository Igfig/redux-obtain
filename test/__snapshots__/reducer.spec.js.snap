// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer Add Resource should add a multiple resources 1`] = `
Object {
  "first_resource_name": Object {
    "data": Object {},
    "error": false,
    "loading": false,
    "paginationKey": undefined,
  },
  "second_resource_name": Object {
    "data": Object {},
    "error": false,
    "loading": false,
    "paginationKey": undefined,
  },
}
`;

exports[`Reducer Add Resource should add a pagination key 1`] = `
Object {
  "paginated_resource": Object {
    "data": Object {},
    "error": false,
    "loading": false,
    "paginationKey": "name_of_attribute_to_be_paginated_on",
  },
}
`;

exports[`Reducer Add Resource should add a resource 1`] = `
Object {
  "new_resource": Object {
    "data": Object {},
    "error": false,
    "loading": false,
    "paginationKey": undefined,
  },
}
`;

exports[`Reducer Fetch Additional Success should append new data onto paginationKey 1`] = `
Object {
  "resource name": Object {
    "data": Object {
      "pagination key": Array [
        1,
        2,
        3,
        4,
        5,
        6,
        7,
        8,
      ],
    },
    "error": false,
    "loading": false,
    "paginationKey": "pagination key",
  },
}
`;

exports[`Reducer Fetch Additional Success should apply accept response to new data 1`] = `
Object {
  "resource name": Object {
    "data": Object {
      "pagination key": Array [
        1,
        2,
        3,
        4,
        10,
        12,
        14,
        16,
      ],
      "plus": "extra",
    },
    "error": false,
    "loading": false,
    "paginationKey": "pagination key",
  },
}
`;

exports[`Reducer Fetch Additional Success should do nothing if the resource does not exist 1`] = `Object {}`;

exports[`Reducer Fetch Additional Success should set the flags correctly 1`] = `
Object {
  "eh whatever": Object {
    "data": Object {
      "pKey": Array [
        1,
        2,
        3,
      ],
    },
    "error": false,
    "loading": false,
    "paginationKey": "pKey",
  },
}
`;

exports[`Reducer Fetch Error should do nothing if resource does not exist 1`] = `Object {}`;

exports[`Reducer Fetch Error should set flags and clear data 1`] = `
Object {
  "unfortunate resource": Object {
    "data": Object {},
    "error": true,
    "loading": false,
    "paginationKey": undefined,
  },
}
`;

exports[`Reducer Fetch Success should apply acceptResponse to data 1`] = `
Object {
  "list": Array [
    1,
    2,
    3,
    4,
    5,
    66,
  ],
  "object": Object {
    "foo": "baz",
  },
}
`;

exports[`Reducer Fetch Success should do nothing if the resource doesn't exist 1`] = `Object {}`;

exports[`Reducer Fetch Success should set data without dataTransform 1`] = `
Object {
  "resource name": Object {
    "data": Object {
      "complicated": Object {
        "nested": Array [
          "structure",
          "and",
          "a",
          "half",
        ],
      },
      "list": Array [
        1,
        2,
        3,
        4,
      ],
    },
    "error": false,
    "loading": false,
    "paginationKey": undefined,
  },
}
`;

exports[`Reducer Fetch Success should set the flags correctly 1`] = `
Object {
  "resource name": Object {
    "data": Object {
      "list": Array [
        1,
        2,
        3,
        4,
      ],
    },
    "error": false,
    "loading": false,
    "paginationKey": undefined,
  },
}
`;

exports[`Reducer Modify Resource should apply dataTransform 1`] = `
Object {
  "whatever": Object {
    "data": Object {
      "newKey": "newValue",
    },
    "error": false,
    "loading": false,
    "paginationKey": undefined,
  },
}
`;

exports[`Reducer Modify Resource should do nothing if no dataTransform is given 1`] = `
Object {
  "whatever": Object {
    "data": Object {},
    "error": false,
    "loading": false,
    "paginationKey": undefined,
  },
}
`;

exports[`Reducer Modify Resource should do nothing if resource does not exist 1`] = `Object {}`;

exports[`Reducer Remove Resource should do nothing if resource does not exist 1`] = `Object {}`;

exports[`Reducer Remove Resource should remove the correct resource 1`] = `
Object {
  "resource_1": Object {
    "data": Object {},
    "error": false,
    "loading": false,
    "paginationKey": undefined,
  },
}
`;

exports[`Reducer Remove Resource should remove the correct resource 2`] = `Object {}`;

exports[`Reducer Request Resource should do nothing if the resource does not exist 1`] = `Object {}`;

exports[`Reducer Request Resource should set flags on request 1`] = `
Object {
  "random name": Object {
    "data": Object {},
    "error": false,
    "loading": true,
    "paginationKey": undefined,
  },
}
`;

exports[`Reducer init should return initial state 1`] = `Object {}`;
